src/pynwb/image.py

```diff
@@ -73,19 +79,104 @@ def __init__(self, **kwargs):
         if unit is None:
             kwargs['unit'] = ImageSeries.DEFAULT_UNIT
 
+        if (
+            args_to_set["external_file"] is not None
+            and args_to_set["starting_frame"] is None
+        ):
+            args_to_set["starting_frame"] = (
+                [0] if len(args_to_set["external_file"]) == 1 else None
+            )
+
         super().__init__(**kwargs)
 
-        if args_to_set["external_file"] is None:
-            args_to_set["starting_frame"] = None  # overwrite starting_frame
         for key, val in args_to_set.items():
             setattr(self, key, val)
 
+        if not self._check_external_file_starting_frame_length():
+            msg = ("%s '%s': The number of frame indices in 'starting_frame' should have "
+                   "the same length as 'external_file'." % (self.__class__.__name__, name))
+            self._raise_error_when_check_is_violated(error_msg=msg)
+
+        if self._change_external_file_format():
+            warnings.warn(
+                "%s '%s': The value for 'format' has been changed to 'external'. "
+                "Setting a default value for 'format' is deprecated and will be changed "
+                "to raising a ValueError in the next release."
+                % (self.__class__.__name__, self.name),
+                DeprecationWarning,
+            )
+
+        if not self._check_external_file_format():
+            msg = (
+                "%s '%s': Format must be 'external' when external_file is specified."
+                % (self.__class__.__name__, name)
+            )
+            self._raise_error_when_check_is_violated(error_msg=msg)
+
+        if not self._check_external_file_data():
+            msg = (
+                "%s '%s': Either external_file or data must be specified (not None), but not both."
+                % (self.__class__.__name__, name)
+            )
+            self._raise_error_when_check_is_violated(error_msg=msg)
+
         if not self._check_image_series_dimension():
             warnings.warn(
                 "%s '%s': Length of data does not match length of timestamps. Your data may be transposed. "
-                "Time should be on the 0th dimension" % (self.__class__.__name__, self.name)
+                "Time should be on the 0th dimension"
+                % (self.__class__.__name__, self.name)
             )
 
+    def _raise_error_when_check_is_violated(self, error_msg):
+        """Raise an error when a check is violated on instance creation.
+        To ensure backwards compatibility, this method throws a warning
+        instead of raising an error when reading from a file."""
+        if not self._in_construct_mode:
+            raise ValueError(error_msg)
+        warnings.warn(error_msg)
```

Comment:
It would be nice to have also a test for Line 136 to check that warning is raised on read. Creating invalid files to test all these options may be too involved just to test this behavior, but I think we could just copy the pattern from the ObjectMapper itself in the test:

https://github.com/hdmf-dev/hdmf/blob/7be40229d778dea940134753de957f2457a11e1f/src/hdmf/build/objectmapper.py#L1257-L1262

I.e., call ``__new__`` explicitly to set ``in_construct_mode=True`` and then call ``__init__`` with the invalid data to check if the warnings are raised as expected. I.e., a test would look something like:

```
def test_warn_on_bad_starting_frame():
     obj = ImageSeries.__new__(ImageSeries, container_source=None, parent=None, object_id="test", in_construct_mode=True)
      with self.assertWarnsWith(ValueError,  "...message..."):
           obj.__init__(........arguments for ImageSeries.....)
     obj._in_construct_mode = False 
```

Setting ``obj._in_construct_mode = False`` should only be necessary of you continue to the use the object.

Reply:
Ok, looking at the tests, it looks like you are already testing this with actual files. Maybe the backward compatibility tests are not counted in codecov? @rly

@weiglszonja so I don't think we need to add more tests, but it would still be interesting to know why codecov indicates that Line 136 ids not covered when it seems like it should be based on the tests you added.

Reply:
Backward compatibility tests are currently not counted in codecov, but now that we are including integration tests in codecov, I think it makes sense to include the backward compatibility tests too.

Reply:
Thanks @rly for the clarification. That makes sense. I added a couple of extra unit-tests anyways to cover this case there as well. I think those tests are useful also because they show how we can write unittests that mimic the use-case were we need to distinguish between construct mode on read and regular user mode.